man/hp6632.1.in
.TH hp6632 1
.SH NAME
hp6632 \- control HP 6632A/6633A/6634A power supplies
.SH SYNOPSIS
.nf
.B hp6632 [\fIOPTIONS\fR]
.fi
.SH DESCRIPTION
\fBhp6632\fR controls the HP 6632A/6633A/6634A power supply via its 
GPIB interface.
\fBhp6632\fR performs actions in the order they appear on the command line.
.SH OPTIONS
\fBhp6632\fR accepts the following command line options:
.TP
\fB\-n\fR, \fB\-\-name\fR name
Set the name used to look up the instrument address in
the config file.  The default is \fIhp6632\fR.
.TP
\fB\-a\fR, \fB\-\-address\fR address
Set the instrument address.
See gpib-utils.conf(5) for a description of address formats.
.TP
\fB\-c\fR, \fB\-\-clear\fR
Clear the instrument to its power-on state, disables its output,
and clears any OVP or OCP condition.
.TP
\fB\-l\fR, \fB\-\-local\fR
Enable the front panel.
.TP
\fB\-i\fR, \fB\-\-get-idn\fR 
Print instrument model.
.TP
\fB\-R\fR, \fB\-\-get-rom\fR 
Print ROM version.
.TP
\fB\-S\fR, \fB\-\-selftest\fR 
Run instrument selftest.
.TP
\fB\-I\fR, \fB\-\-iset\fR float
Set current in amps.
.TP
\fB\-V\fR, \fB\-\-vset\fR float
Set voltage in volts.
.TP
\fB\-o\fR, \fB\-\-out\fR 0|1
Enable/disable output.
.TP
\fB\-O\fR, \fB\-\-ovset\fR volts
Set the over-voltage threshold in volts.
If the output ever exceeds this value, the OVP circuit shorts the output
and disables the supply.
To clear this condition, use the \fB\-\-clear\fR option.
.TP
\fB\-C\fR, \fB\-\-ocp\fR 0|1
Enable/disable overcurrent protection.
If OCP is enabled and output mode changes from CV (constant voltage)
to CC (constant current mode), the OCP trips and disables the output.
To clear this condition, use the \fB\-\-clear\fR option.
.TP
\fB\-s\fR, \fB\-\-samples\fR number
Select number of current and voltage samples to take [Default 0].
.TP
\fB\-p\fR, \fB\-\-period\fR period
Set sample period [Default 0s].
.SH FILES
@X_SYSCONFDIR@/gpib-utils.conf
.br
~/.gpib-utils.conf
.SH "SEE ALSO"
gpib-utils.conf(5)
.PP
"Operating Manual, HP-IB System DC Power Supplies, HP Models 6632A, 
6633A, 6634A", May 1992.


NEWS
=========================================================================
gpib-utils-1.5 Release Notes                                   2014-08-21
=========================================================================

Moving code from sourceforge to github:

* Stop tracking autotools build products, etc.

* Remove ChangeLog (use commit log and NEWS), add README.md

These changes were staged in 2011:

* Export libvxi11.so library for general use.

* libvxi11 API cleanup, portability fix to gpib.c [Alessio Bogani].

* New utilities
  - ibquery:  Generic query utility.

=========================================================================
gpib-utils-1.4 Release Notes                                   2009-02-02 
=========================================================================

* Added support for new devices:
  - ics8064:  ICS Electronics 8064 Ethernet to Relay interface
  - a34401:   Agilent 34401A/34410A/34411A DMM.
  - hp6038:   Hewlett-Packard 603xA System DC Power Supply
  - hp6626:   Hewlett-Packard 6625A/6626A/6628A/6629A System DC Power Supply
  - hp6632:   Hewlett-Packard 6632A/6633A/6634A System DC Power Supply

* Handle non-zero GPIB board index and secondary GPIB addresses.

* Changed ics8065 utility to more general purpose icsconfig utility.

* Added automake/autoconf infrastructure so gpib-utils compiles with and
  without linux-gpib.

* Refactored ICS, VXI-11, and LSD code into (non-installed) libraries.

=========================================================================
gpib-utils-1.3 Release Notes                                   2007-11-23 
=========================================================================

* VXI-11 support for devices behind a Ethernet-GPIB gateway such as
  ICS 8065, or with built-in Ethernet ports such as the Agilent 6000 
  series oscilloscopes.

* Added config file: /etc/gpib-utils.conf.  Place device addresses here
  and overridde with -a option on command line.  If linux-gpib is used,
  /etc/gpib.conf need only contain bus master device information.

* Added support for new devices:
  - a6032 Agilent 6000-series oscilloscopes
  - dg535 Stanford Research Systems DG535 digital delay/pulse generator 
  - sr620 Stanford Research Systems SR620 Universal Time Interval Counter
  - ics8065 ICS Electroncs 8065 Ethernet-GPIB Controller
  - hp3488 Hewlett-Packard 3488A Switch/Control Unit

* Removed support for old instruments that I no longer have access to
  and which I know of nobody using:
  - r5005 Racal Dana model 5005 DMM
  - hp3455 Hewlett-Packard 3455A DMM
  - tek2230 Tektronix 2230 oscilloscope (never really working)


src/hp6632.c
/* This file is part of gpib-utils.
   For details, see http://sourceforge.net/projects/gpib-utils.

   Copyright (C) 2008-2009 Jim Garlick <garlick.jim@gmail.com>

   gpib-utils is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   gpib-utils is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with gpib-utils; if not, write to the Free Software Foundation, 
   Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA */

/* HP 6632A (25V, 4A); 6633A (50V, 2A); 6634A (100V, 1A) power supplies */

#if HAVE_CONFIG_H
#include "config.h"
#endif
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <errno.h>
#include <string.h>
#include <unistd.h>
#include <libgen.h>
#if HAVE_GETOPT_LONG
#include <getopt.h>
#endif
#include <assert.h>
#include <sys/time.h>
#include <math.h>
#include <stdint.h>

#include "libutil/util.h"
#include "gpib.h"

#define INSTRUMENT "hp6632"

/* serial poll reg */
#define HP6632_SPOLL_FAU    1
#define HP6632_SPOLL_PON    2
#define HP6632_SPOLL_RDY    16
#define HP6632_SPOLL_ERR    32
#define HP6632_SPOLL_RQS    64

/* status reg. bits (STS? - also ASTS? UNMASK? FAULT?) */
#define HP6632_STAT_CV      1       /* constant voltage mode */
#define HP6632_STAT_CC      2       /* positive constant current mode */
#define HP6632_STAT_UNR     4       /* output unregulated */
#define HP6632_STAT_OV      8       /* overvoltage protection tripped */
#define HP6632_STAT_OT      16      /* over-temperature protection tripped */
#define HP6632_STAT_OC      64      /* overcurrent protection tripped */
#define HP6632_STAT_ERR     128     /* programming error */
#define HP6632_STAT_INH     256     /* remote inhibit */
#define HP6632_STAT_NCC     512     /* negative constant current mode */
#define HP6632_STAT_FAST    1024    /* output in fast operating mode */
#define HP6632_STAT_NORM    2048    /* output in normal operating mode */

/* programming errors */
static strtab_t petab[] = {
    { 0, "Success" },
    { 1, "EEPROM save failed" },
    { 2, "Second PON after power-on" },
    { 4, "Second DC PON after power-on" },
    { 5, "No relay option present" },
    { 8, "Addressed to talk and nothing to say" },
    { 10, "Header expected" },
    { 11, "Unrecognized header" },
    { 20, "Number expected" },
    { 21, "Number syntax" },
    { 22, "Number out of internal range" },
    { 30, "Comma expected" },
    { 31, "Terminator expected" },
    { 41, "Parameter out of range" },
    { 42, "Voltage programming error" },
    { 43, "Current programming error" },
    { 44, "Overvoltage programming error" },
    { 45, "Delay programming error" },
    { 46, "Mask programming error" },
    { 50, "Multiple CSAVE commands" },
    { 51, "EEPROM checksum failure" },
    { 52, "Calibration mode disabled" },
    { 53, "CAL channel out of range" },
    { 54, "CAL FS out of range" },
    { 54, "CAL offset out of range" },
    { 54, "CAL disable jumper in" },
    { 0, NULL }
};

/* selftest failures */
static strtab_t sttab[] = {
    { 0, "Passed" },
    { 1, "ROM checksum failure" },
    { 2, "RAM test failure" },
    { 3, "HP-IB chip failure" },
    { 4, "HP-IB microprocessor timer slow" },
    { 5, "HP-IB microprocessor timer fast" },
    { 11, "PSI ROM checksum failure" },
    { 12, "PSI RAM test failure" },
    { 14, "PSI microprocessor timer slow" },
    { 15, "PSI microprocessor timer fast" },
    { 16, "A/D test reads high" },
    { 17, "A/D test reads low" },
    { 18, "CV/CC zero too high" },
    { 19, "CV/CC zero too low" },
    { 20, "CV ref FS too high" },
    { 21, "CV ref FS too low" },
    { 22, "CC ref FS too high" },
    { 23, "CC ref FS too low" },
    { 24, "DAC test failed" },
    { 51, "EEPROM checksum failed" },
    { 0, NULL }
};

static int _interpret_status(gd_t gd, unsigned char status, char *msg);
static int _readiv(gd_t gd, double *ip, double *vp);

char *prog = "";

#define OPTIONS OPTIONS_COMMON "cliSI:V:o:O:C:Rs:p:"
#if HAVE_GETOPT_LONG
#define GETOPT(ac,av,opt,lopt) getopt_long(ac,av,opt,lopt,NULL)
static struct option longopts[] = {
    OPTIONS_COMMON_LONG,
    {"verbose",         no_argument,       0, 'v'},
    {"clear",           no_argument,       0, 'c'},
    {"local",           no_argument,       0, 'l'},
    {"get-idn",         no_argument,       0, 'i'},
    {"get-rom",         no_argument,       0, 'R'},
    {"selftest",        no_argument,       0, 'S'},
    {"iset",            required_argument, 0, 'I'},
    {"vset",            required_argument, 0, 'V'},
    {"out",             required_argument, 0, 'o'},
    {"ovset",           required_argument, 0, 'O'},
    {"ocp",             required_argument, 0, 'C'},
    {"samples",         required_argument, 0, 's'},
    {"period",          required_argument, 0, 'p'},
    {0, 0, 0, 0},
};
#else
#define GETOPT(ac,av,opt,lopt) getopt(ac,av,opt)
#endif

static opt_desc_t optdesc[] = {
    OPTIONS_COMMON_DESC,
    { "c", "clear",   "initialize instrument to default values" },
    { "l","local",    "return instrument to local operation on exit" },
    { "i","get-idn",  "print instrument model" },
    { "R","get-rom",  "print instrument ROM version" },
    { "S","selftest", "run selftest" },
    { "I","iset",     "set current" },
    { "V","vset",     "set voltage" },
    { "o","out",      "enable/disable output (0|1)" },
    { "O","ovset",    "set overvoltage threshold" },
    { "C","ocp",      "enable/disable overcurrent protection (0|1)" },
    { "s","samples",  "number of samples [0]" },
    { "p","period",   "sample period" },
    {0, 0},
};

int
main(int argc, char *argv[])
{
    gd_t gd;
    int rc, c, print_usage = 0;
    int exit_val = 0;
    char tmpstr[64];
    int samples = 0;
    double period = 0.0;
    int showtime = 0;

    gd = gpib_init_args(argc, argv, OPTIONS, longopts, INSTRUMENT,
                        _interpret_status, 0, &print_usage);
    if (print_usage)
        usage(optdesc);
    if (!gd)
        exit(1);

    while ((c = GETOPT(argc, argv, OPTIONS, longopts)) != EOF) {
        switch (c) {
        OPTIONS_COMMON_SWITCH
            break;
        case 'c': /* --clear */
            gpib_clr(gd, 1000000); /* same as 'CLR' command */
            gpib_wrtf(gd, "OUT 0;RST\n");
            sleep(1);
            break;
        case 'l': /* --local */
            gpib_loc(gd); 
            break;
        case 'i': /* --get-idn */
            gpib_wrtstr(gd, "ID?\n");
            gpib_rdstr(gd, tmpstr, sizeof(tmpstr));
            printf("%s\n", tmpstr);
            break;
        case 'R': /* --get-rom */
            gpib_wrtstr(gd, "ROM?\n");
            gpib_rdstr(gd, tmpstr, sizeof(tmpstr));
            printf("rom version: %s\n", tmpstr);
            break;
        case 'S': /* --selftest */
            gpib_wrtstr(gd, "TEST?\n");
            gpib_rdstr(gd, tmpstr, sizeof(tmpstr));
            if (sscanf(tmpstr, "%d", &rc) == 1)
                printf("self-test: %s\n", findstr(sttab ,rc));
            else
                printf("self-test: %s\n", tmpstr);
            break;
        case 'I': /* --iset */
            gpib_wrtf(gd, "ISET %s\n", optarg);
            break;
        case 'V': /* --vset */
            gpib_wrtf(gd, "VSET %s\n", optarg);
            break;
        case 'O': /* --ovset */
            gpib_wrtf(gd, "OVSET %s\n", optarg);
            break;
        case 'o': /* --out */
            gpib_wrtf(gd, "OUT %s\n", optarg);
            break;
        case 'C': /* --ocp */
            gpib_wrtf(gd, "OCP %s\n", optarg);
            break;
        case 's': /* --samples */
            samples = strtoul(optarg, NULL, 10);
            if (samples > 1)
                showtime = 1;
            break;
        case 'p': /* --period */
            if (freqstr(optarg, &period) < 0) {
                fprintf(stderr, "%s: error parsing period argument\n", prog);
                fprintf(stderr, "%s: use units of: %s\n", prog, PERIOD_UNITS);
                exit_val = 1;
                break;
            }
            period = 1.0/period;
            break;
        }
    }

    if (samples > 0) {
        double t0, t1, t2;
        double i, v;

        t0 = 0;
        while (samples-- > 0) {
            t1 = gettime();
            if (t0 == 0)
                t0 = t1;
            if (_readiv(gd, &i, &v) == -1) {
                exit_val = 1;
                goto done;  
            }
            t2 = gettime();

            if (showtime)
                printf("%-3.3lf\t%lf\t%lf\n", (t1-t0), i, v);
            else
                printf("%lf\t%lf\n", i, v);
            if (samples > 0 && period > 0)
                sleep_sec(period - (t2-t1));
        }
    }

done:
    gpib_fini(gd);
    exit(exit_val);
}


static int
_interpret_status(gd_t gd, unsigned char status, char *msg)
{
    char tmpstr[64];    
    int e, err = 0;

    if (status & HP6632_SPOLL_FAU) {
        fprintf(stderr, "%s: device fault\n", prog);
        /* FIXME: check FAULT? and/or ASTS? but only if unmasked */
        err = 1;
    }
    if (status & HP6632_SPOLL_PON) {
        /*fprintf(stderr, "%s: power-on detected\n", prog);*/
    }
    if (status & HP6632_SPOLL_RDY) {
    }
    if (status & HP6632_SPOLL_ERR) {
        gpib_wrtstr(gd, "ERR?\n");
        gpib_rdstr(gd, tmpstr, sizeof(tmpstr));
        if (sscanf(tmpstr, "%d", &e) == 1)
            fprintf(stderr, "%s: prog error: %s\n", prog, findstr(petab, e));
        else
            fprintf(stderr, "%s: prog error: %s\n", prog, tmpstr);
        err = 1;
    }
    if (status & HP6632_SPOLL_RQS) {
        fprintf(stderr, "%s: device is requesting service\n", prog);
        err = 1; /* it shouldn't be unless we tell it to */
    }
    return err;
}

static int 
_readiv(gd_t gd, double *ip, double *vp)
{
    gpib_wrtstr(gd, "VOUT?\n");
    if (gpib_rdf(gd, "%lf", vp) != 1) {
        fprintf(stderr, "%s: error reading VOUT? result\n", prog);
        return -1;
    }
    gpib_wrtstr(gd, "IOUT?\n");
    if (gpib_rdf(gd, "%lf", ip) != 1) {
        fprintf(stderr, "%s: error reading IOUT? result\n", prog);
        return -1;
    }
    return 0;
}

/*
 * vi:tabstop=4 shiftwidth=4 expandtab
 */
